File: src/git.ts
import { execFileSync } from "node:child_process";
import { resolve, join } from "node:path";
import { statSync, readFileSync } from "node:fs";
import { FileDiff } from "./schema.js";
import { DEFAULTS } from "./config.js";

const BINARY_EXTENSIONS = new Set([
  ".png", ".jpg", ".jpeg", ".gif", ".webp", ".ico", ".svg",
  ".pdf", ".zip", ".tar", ".gz", ".bz2", ".xz", ".whl",
  ".pyc", ".pyo", ".so", ".dylib", ".dll", ".exe",
  ".db", ".sqlite", ".sqlite3",
]);

export function getChangedFiles(
  baseRef: string,
  headRef: string = "HEAD",
  repoRoot: string = "."
): FileDiff[] {
  const root = resolve(repoRoot);

  const nameStatus = runGit(["diff", "--name-status", baseRef, headRef], root);

  const results: FileDiff[] = [];

  for (const line of nameStatus.split("\n")) {
    if (!line.trim()) continue;

    const parts = line.split("\t");
    const status = (parts[0] ?? "").trim();

    let filePath: string;
    if (status.startsWith("R") && parts.length >= 3) {
      filePath = parts[2]!;
    } else {
      filePath = parts[1]!;
    }

    const isDeleted = status === "D";
    const isNew = status === "A";

    if (isBinary(filePath)) {
      results.push({
        path: filePath,
        diff: "",
        content: null,
        is_binary: true,
        is_deleted: isDeleted,
        is_new: isNew,
      });
      continue;
    }

    let diffOutput: string;
    try {
      diffOutput = runGit(["diff", baseRef, headRef, "--", filePath], root);
    } catch {
      diffOutput = "";
    }

    if (diffOutput.includes("Binary files")) {
      results.push({
        path: filePath,
        diff: "",
        content: null,
        is_binary: true,
        is_deleted: isDeleted,
        is_new: isNew,
      });
      continue;
    }

    let content: string | null = null;
    if (!isDeleted) {
      try {
        content = runGit(["show", `${headRef}:${filePath}`], root);
        if (Buffer.byteLength(content, "utf-8") > DEFAULTS.maxFileBytes) {
          content = null;
        }
      } catch {
        const absPath = join(root, filePath);
        content = getFileContent(absPath);
      }
    }

    results.push({
      path: filePath,
      diff: diffOutput,
      content,
      is_binary: false,
      is_deleted: isDeleted,
      is_new: isNew,
    });
  }

  return results;
}

function isBinary(filePath: string): boolean {
  const ext = filePath.slice(filePath.lastIndexOf(".")).toLowerCase();
  return BINARY_EXTENSIONS.has(ext);
}

export function getFileContent(absPath: string): string | null {
  try {
    const stat = statSync(absPath);
    if (stat.size > DEFAULTS.maxFileBytes) return null;
    return readFileSync(absPath, { encoding: "utf-8" });
  } catch {
    return null;
  }
}

function runGit(args: string[], cwd: string): string {
  try {
    return execFileSync("git", args, { cwd, encoding: "utf-8" });
  } catch (err) {
    const e = err as { stderr?: string; message?: string };
    throw new Error(`git ${args.join(" ")} failed: ${e.stderr?.trim() ?? e.message}`);
  }
}

/** Read a file from a specific git ref. Returns null if the ref or path does not exist. */
export function getFileAtRef(relPath: string, ref: string, repoRoot: string): string | null {
  try {
    return runGit(["show", `${ref}:${relPath}`], resolve(repoRoot));
  } catch {
    return null;
  }
}

File: rust/src/git.rs
//! Git operations: run commands, get changed files, show file content
//!
//! Uses std::process::Command (blocking). Wrap in spawn_blocking if needed.

use anyhow::{bail, Context, Result};
use std::path::Path;

use crate::schema::FileDiff;

/// List of binary file extensions to skip
const BINARY_EXTENSIONS: &[&str] = &[
    "png", "jpg", "jpeg", "gif", "bmp", "ico", "webp", "svg", "pdf", "doc", "docx", "xls", "xlsx",
    "ppt", "pptx", "zip", "tar", "gz", "bz2", "7z", "rar", "exe", "dll", "so", "dylib", "bin",
    "ttf", "otf", "woff", "woff2", "eot", "mp3", "mp4", "avi", "mov", "mkv", "webm", "lock",
    "lockb",
];

/// Run a git command and return stdout
pub fn run_git(args: &[&str], cwd: &Path) -> Result<String> {
    let output = std::process::Command::new("git")
        .args(args)
        .current_dir(cwd)
        .output()
        .with_context(|| format!("failed to run: git {}", args.join(" ")))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        bail!("git {} failed: {}", args.join(" "), stderr.trim());
    }

    Ok(String::from_utf8_lossy(&output.stdout).into_owned())
}

/// Run git command, returning None on failure instead of error
pub fn run_git_optional(args: &[&str], cwd: &Path) -> Option<String> {
    run_git(args, cwd).ok()
}

/// Get the repository root directory
pub fn get_repo_root(cwd: &Path) -> Result<std::path::PathBuf> {
    let output = run_git(&["rev-parse", "--show-toplevel"], cwd)?;
    Ok(std::path::PathBuf::from(output.trim()))
}

/// Check if a file path has a binary extension
pub fn is_binary_extension(path: &str) -> bool {
    path.rsplit('.')
        .next()
        .map(|ext| BINARY_EXTENSIONS.contains(&ext.to_lowercase().as_str()))
        .unwrap_or(false)
}

/// Get list of changed files between two refs, enriched with diff and content
pub fn get_changed_files(
    base_ref: &str,
    head_ref: &str,
    repo_root: &Path,
    max_file_bytes: u64,
) -> Result<Vec<FileDiff>> {
    let output = run_git(&["diff", "--name-status", base_ref, head_ref], repo_root)?;

    let mut files = Vec::new();
    for line in output.lines() {
        if line.trim().is_empty() {
            continue;
        }

        let parts: Vec<&str> = line.split('\t').collect();
        if parts.is_empty() {
            continue;
        }

        let status_char = parts[0].chars().next().unwrap_or('M');
        let is_deleted = status_char == 'D';
        let is_new = status_char == 'A';
        let is_renamed = status_char == 'R';

        // For renamed files, use the new name (second path)
        let path = if is_renamed && parts.len() >= 3 {
            parts[2].to_string()
        } else if parts.len() >= 2 {
            parts[1].to_string()
        } else {
            continue;
        };

        let diff =
            run_git(&["diff", base_ref, head_ref, "--", &path], repo_root).unwrap_or_default();

        let is_binary = is_binary_extension(&path) || diff.contains("Binary files");

        let (content, is_oversized, oversized_bytes) = if !is_deleted && !is_binary {
            let spec = format!("{}:{}", head_ref, path);
            match run_git_optional(&["show", &spec], repo_root) {
                Some(c) => {
                    let byte_len = c.len() as u64;
                    if byte_len > max_file_bytes {
                        (None, true, Some(byte_len))
                    } else {
                        (Some(c), false, None)
                    }
                }
                None => (None, false, None),
            }
        } else {
            (None, false, None)
        };

        files.push(FileDiff {
            path,
            diff,
            content,
            is_binary,
            is_deleted,
            is_new,
            is_oversized,
            oversized_bytes,
        });
    }

    Ok(files)
}

/// Read a local file, returning None if it exceeds max_bytes or is not valid UTF-8
pub fn get_file_content(path: &Path, max_bytes: u64) -> Option<String> {
    let bytes = std::fs::read(path).ok()?;
    if bytes.len() as u64 > max_bytes {
        return None;
    }
    String::from_utf8(bytes).ok()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_binary_extension() {
        assert!(is_binary_extension("image.png"));
        assert!(is_binary_extension("FILE.PNG"));
        assert!(is_binary_extension("path/to/doc.pdf"));
        assert!(!is_binary_extension("code.rs"));
        assert!(!is_binary_extension("Makefile"));
    }
}
